xrootd/client/client.go

```diff
@@ -104,182 +94,77 @@ func NewClient(ctx context.Context, address string, username string, opts ...Opt
+	client.initialSessionID = address
+	client.sessions[client.initialSessionID] = session
 
 	return client, nil
 }
 
 // Close closes the connection. Any blocked operation will be unblocked and return error.
 func (client *Client) Close() error {
-	client.cancel()
-
-	client.mux.Close()
-	return client.conn.Close()
-}
-
-func (client *Client) consume(ctx context.Context) {
-	var header xrdproto.ResponseHeader
-	var headerBytes = make([]byte, xrdproto.ResponseHeaderLength)
-
-	for {
-		select {
-		case <-ctx.Done():
-			// TODO: Should wait for active requests to be completed?
-			return
-		default:
-			if _, err := io.ReadFull(client.conn, headerBytes); err != nil {
-				if ctx.Err() != nil {
-					// something happened to the context.
-					// ignore this error.
-					continue
-				}
-				panic(err)
-				// TODO: handle EOF by redirection as specified at http://xrootd.org/doc/dev45/XRdv310.pdf, page 11
-			}
-
-			if err := xrdproto.Unmarshal(headerBytes, &header); err != nil {
-				if ctx.Err() != nil {
-					// something happened to the context.
-					// ignore this error.
-					continue
-				}
-				panic(err)
-				// TODO: should redirect in case if is not possible to decode a header as well?
-			}
-
-			resp := mux.ServerResponse{Data: make([]byte, header.DataLength)}
-			if _, err := io.ReadFull(client.conn, resp.Data); err != nil {
-				if ctx.Err() != nil {
-					// something happened to the context.
-					// ignore this error.
-					continue
-				}
-				resp.Err = err
-			} else if header.Status != xrdproto.Ok {
-				resp.Err = header.Error(resp.Data)
-			}
-
-			if err := client.mux.SendData(header.StreamID, resp); err != nil {
-				if ctx.Err() != nil {
-					// something happened to the context.
-					// ignore this error.
-					continue
-				}
-				panic(err)
-				// TODO: should we just ignore responses to unclaimed stream IDs?
-			}
-
-			if header.Status != xrdproto.OkSoFar {
-				client.mux.Unclaim(header.StreamID)
-			}
-		}
-	}
+	defer client.cancel()
+	// TODO: cleanup sessions.
+	return nil
 }
 
 // Send sends the request to the server and stores the response inside the resp.
-func (client *Client) Send(ctx context.Context, resp xrdproto.Response, req xrdproto.Request) error {
-	data, err := client.call(ctx, req)
-	if err != nil {
-		return err
-	}
-
-	return resp.UnmarshalXrd(xrdenc.NewRBuffer(data))
+// If the resp is nil, then no response is stored.
+func (client *Client) Send(ctx context.Context, resp xrdproto.Response, req xrdproto.Request) (string, error) {
+	return client.SendSession(ctx, client.initialSessionID, resp, req)
 }
 
-func (client *Client) send(ctx context.Context, responseChannel mux.DataRecvChan, request []byte) ([]byte, error) {
-	if _, err := client.conn.Write(request); err != nil {
-		return nil, err
+func (client *Client) SendSession(ctx context.Context, sessionID string, resp xrdproto.Response, req xrdproto.Request) (string, error) {
```

Comment:
do we need to export this method?
it seems to me it can be kept internal. (ie: `s/SendSession/sendSession/`)